src/eos-companion-app-integration-helper.c

```diff
@@ -127,25 +122,25 @@ eos_companion_app_service_soup_server_listen_on_sd_fd_or_port (SoupServer
                                                                SoupServerListenOptions   options,
                                                                GError                  **error)
 {
-  g_autoptr(GError) local_error = NULL;
+  const char *sd_fd_value = g_getenv ("EOS_COMPANION_APP_SERVICE_LISTEN_FD");
```

Comment:
If I understand the issue correctly, the problem is not that there could ever be a file descriptor different than 3 passed by systemd, or anything else for that matter, as we would never use the file descriptor code path when running from the terminal.
The issue is that the try/recover mechanism we use now can't be used to reliably detect whether we're launched from systemd.

In other words there should be no reason to convert the env var value string to int; we could still hardcode `3` and add an environment variable in the service file (if systemd does not set one already) that simply signifies that we were started by systemd.

Reply:
> In other words there should be no reason to convert the env var value string to int; we could still hardcode 3 and add an environment variable in the service file (if systemd does not set one already) that simply signifies that we were started by systemd.

True. I also considered this approach, though just setting an environment variable with the fd seemed easier.

Feel free to push a follow up and merge.